add rotation policies concepts to 2.x (#2090)
hashicorp/web-unified-docs: content/vault/v2.x (rc)/content/docs/concepts/rotation-policies.mdx

---
layout: docs
page_title: Rotation Policies
description: >-
  Rotation policies allow administrators to define tailored retry logic and determine how to

>   process failing rotation entries in the Rotation Manager.

Reviewer: ```suggestion
  Rotation policies let you define tailored retry logic and tell Vault how to
  process failing rotation entries in the rotation manager.
```
Style correction: reserve capitalization for product names, speak directly to the reader

---

# Rotation policies

> @include 'alerts/enterprise-only.mdx'

Reviewer: ```suggestion

@include 'alerts/enterprise-only.mdx'
```
The markdown parser gets fussy when includes don't have whitespace before/after the include line

Rotation Policies are a set of instructions used by the Rotation Manager to define standardized
error-handling behaviors and lifecycle tracking for onboarded credentials. Rotation Policies

> are managed via the Vault API.

Reviewer: ```suggestion
Vault includes a rotation manager to enforce standardized error-handling
behaviors and lifecycle tracking for onboarded credentials. You can control how
the rotation manager behaves by providing instructions in the form of rotation
policies.
```
Style correction: write simply and use declarative statements

to fail even after all these attempts, the Rotation Manager will determine that the maximum retry limits
have been exceeded. At this point, the Rotation Manager transitions the credential's rotation to an "Orphaned"
state, halting any further automated rotation attempts.

## Default Rotation policy

Vault ships with a default rotation policy that is automatically used if no rotation policy is specified.
The default rotation policy includes the following parameters:

```json
{
  "max_retry_cycles": 3,
  "max_retries_per_cycle": 6
}
```

## Usage

Rotation policies can be assigned when onboarding a credential to the Rotation Manager.
Below is an example of specifying a rotation policy on an LDAP Secrets Enterprise static role:

```
vault write ldap/static-role/alice \
    dn="cn=alice,ou=users,dc=example,dc=com" \
    username="alice" \
    rotation_period="10s" \
    rotation_policy="custom-ldap-policy"
```

If no policy is specified upon initial configuration, the default rotation policy is used.
The rotation policy can also be updated. In the example below, the initial rotation policy is cleared out.

> This informs the Rotation Manager to use the default rotation policy for this credential going forward:

Reviewer: ```suggestion
If you do not provide a rotation policy when creating a credential, Vault uses
the default policy. You can unset an assigned policy and force Vault to use the 
default policy by updating the credential to unset the `rotation_policy` field:
```

```
vault write ldap/static-role/alice \
    dn="cn=alice,ou=users,dc=example,dc=com" \
    username="alice" \
    rotation_period="10s" \
    rotation_policy=""
```

Rotation Manager fetches the latest policy updates on each re-queue cycle. Note that policy updates

> may not immediately be reflected.

Reviewer: ```suggestion
The rotation manager fetches the latest policy updates on each requeue cycle,
which means you may not see policy updates reflected immediately in Vault behvior.
```

- Doubling Interval: Subsequent delays double the previous backoff interval (e.g., 10s, 20s, 40s, 80s, etc.).
- Maximum Cap: The delay for any single retry is capped at a maximum of 5 minutes (300 seconds).
- Jitter: A randomized factor (jitter) is applied to the calculated delay to ensure more even load distribution
  across the cluster.

The `max_retries_per_cycle` parameter defines the explicit limit for how many times this exponential backoff
schedule is applied during a single rotation cycle.

## Orphaning

When a rotation entry exceeds the maximum retry limits defined by its policy, the Rotation Manager
transitions the credential to an **Orphaned** state, halting all further automated rotation attempts.
This mechanism is used to prevent the Rotation Manager from being clogged by persistent errors.

Restoring an orphaned rotation entry requires manual intervention from the plugin administrator.

To re-enable automated rotations on a credential:
- resolve the underlying error that caused the credential to be orphaned.
- manually re-register the entry with the Rotation Manager by re-running its original configuration command.

>   This action re-registers the entry without changing any configuration data.

Reviewer: ```suggestion
1. Update the credential to reassign the rotation policy and register the entry
   with the rotation manager. You can use the same original configuration
   command to reregister the entry without changing the other configuration data.
```
